### ISSUE TITLE
user interface design

### ISSUE THREAD

**Issue by jeleniste** [VasiliosKalogirou/MongoDB-QGIS#2]

**Comment by jeleniste** [comment]
I added my ui design to new folder in this repo. Feel free to change it. It can be opened from qt4designer.

**Comment by VasiliosKalogirou** [comment]
Great, thanks Jan! Will do tomorrow!

**Comment by jeleniste** [comment]
@VasiliosKalogirou it is only ui from my plugin.

**Comment by VasiliosKalogirou** [comment]
@jeleniste added the dialog for storing data to the interface. Made a few changes (capital letters) to all the dialogs so that they can match to each other. Also, changed the name of the interface as "MongoDB-Manager". Didn't know that to put in there so I got inspiration from the "DB Manager" plugin. Please have a look and let me know your thoughts, if you like it, then we can start adding code!

**Comment by jeleniste** [comment]
on your card has controls another align, i will fix it.

**Comment by VasiliosKalogirou** [comment]
Sure. How should we go on and add the code from now? Should we create another plugin using Plugin-builder and place this .ui file in the folder?

**Comment by mk45** [comment]
1. I would add field / info "currently connected to:" that way user can be connected by default/automatic to localhost:27017 (if exists) and change it and/or reconnect if wish (that is I believe common action so do we force user to click? )
   So is info Label on top of Connection tab ok? I'll do it.
2. Load Tab. shouldn't there be "Load" instead of "run" on button? I'll do it.
3. I don't understand save tab. Why people can't input collection and database by LineEdit widget? What if collection already exists? It should not be used at all or User must agree directly for cleaning previous.
4. Longer I look at tabs more I believe all tabs should have Help in label (maybe on top). with  just brief information what's the tab-action result (As a user I have no Idea what "Try"  does)
5. Besides showing created dialog/tabs to user  there are also a top menu (database menu) and toolbar (database toolbar). Which leads to another questions are there be any other actions in menu than just show dialog ( we may think about "save layer" as action/icon in menu or toolbar and so on...). Finally what about icon do @jeleniste gives his nice one?
6. What language do We write this in? C - I'm out . Python - You can take my plugin init. But we need to talk about License. GPL is ok for me, but we should talk about it.

**Comment by jeleniste** [comment]
@mk45 
Language: C I`m out too, python is OK, but my python abilities are poor.
License: The same as on other plugins? What license must have plugin for oficial repository.
Icon: We can use mine, but you should know, that this is a small joke, using this shape in czech enviroment. I would like to use mine.

**Comment by mk45** [comment]
I'm  doing my day work using Python. I did some changes in layout to be more form-like and rename buttons. I will add text description later to TextLabels.
By License I mean how do we sing our work (per file, as a team, whoever writes something adds his name to the top, in contributors to project, and so on...). Which one do we choose?

**Comment by VasiliosKalogirou** [comment]
@mk45  What do you mean by point 1? Maybe that as soon as the user runs the plugin he gets connected to localhost by default?
Also, I didn't understand point 3. The "save" tab will be used for users to saving features from a layer in QGIS to MongoDB, if that's what you asking?
I agree with point 4, we could add a "?" next to each label to pop-up a helping message.
I also agree with using Python.
You guys can choose which icon to use, maybe the MongoDB official icon would be nice. 
@jeleniste I don't know much about the license but I guess we could all be contributors to the merged plugin? Maybe we could consult Paolo about this?

**Comment by VasiliosKalogirou** [comment]
@mk45 I like the design! But what are the text labels in each tab? Shouldn't we remove them from the interface or remove the text "TextLabel"?
Also, what do you think about placing a "?" next to each label which will explain what each label is used for?

**Comment by jeleniste** [comment]
@mk45 IMHO placing labels in line of combo/line edits is not goog idea, if we want use this widget as docked on left side.

**Comment by mk45** [comment]
1. I thought about auto connect but now I think it is bad idea.
2. Yes but how to save? Best would be to save laver as new collection or into empty one. That is because no merge should be done in plugin/database level as QGIS merges vector layers well and has tools for that. This leads to question what if collection does not exist (we want to provide new name in edit) or what if is not empty.
3. Tooltip is one thing but expressive annotation what save tab or query tab does as a result, what is database user is connected to, would be nice. User don't reads the documentation nor do further search for help. Those empty TextLabels are for tab description. I think tooltip is better than "?". I remember "?" in top-left in Win95/98 I never used it. 
   I doubt we can use mongo default icon for legal reasons.

@jeleniste You're probably right I'm mot sure what was better. @VasiliosKalogirou What do You think about current layout? If not is ok I revert this. It was just a test. I don't like TextLabels above ..how about description  in LineEdit comment (gray text in field) ?

**Comment by jeleniste** [comment]
@mk45 you can use "vertical spacer"
This is question, if we want have widget docked, or plain, if we want to dock it on left side is better to design it tall/

**Comment by jeleniste** [comment]
@VasiliosKalogirou it should be option for append/drop and recreate existing collection enabled if collection exists on database.

**Comment by jeleniste** [comment]
And we should unify variables and widgets nomenclature if we will use _ or camel notation.

**Comment by VasiliosKalogirou** [comment]
Ok guys, great ideas on the table! :)
1. I totally agree on the idea of unified variables and widgets nomenclature. I suggest  starting with a lower case letter and then following camel notation.
2. I thought about the issue storing data in a collection that does not exist. What about letting the user write the name of the collection and the database (instead of populating a ComboBox with the existing databases and collections) and if the collection of the database does not exist, then a window appears saying: "You have chosen to store your data in a database which does not exist. Do you want to the create the database and proceed with storing your data?" and "You have chosen to store your data in a collection which does not exist. Do you want to the create the collection  and proceed with storing your data?"
3. Regarding designing the interface, personally, I always liked non-docked plugins. I like the interface right now where the label is next to the ComboBox or the LineEdit object. It also makes for a smaller interface. If you want to though, we can revert to the Jan's version and have the plugin docked on the left.
4. I also don't think auto-connect is a good idea. Jan's way of connecting to the database and having this functionality in a separate tab from loading and storing data is brilliant for me!

**Comment by mk45** [comment]
@VasiliosKalogirou so could You make for 1. (rename) and at least sketch of 3. (non-docked version just for general feeling)? 
Who is the project leader ? or do we vote in case of disagreement ?

**Comment by VasiliosKalogirou** [comment]
@mk45 I would recommend @jeleniste as the project leader as he is the once who has done the most work with his plugin. Regarding the docked/undocked debate I prefer undocked. If you guys agree we I can change the design of the interface a little bit, and keep the labels next to the combocoxes and lineEdits as they look like now.

I also have another 2 things to recommend to start with:
1. To add the following global variables at the top of the code: 
host,
port,
username,
password,
authDatabase,
inputDatabase,
inputCollection,
inputGeometryName,
inputSrid,
outputDatabase,
outputCollection,
outputGeometryName,
outputSrid.
2. To add the code for connecting to the database first. This can be done by @jeleniste as this is actually his code. When the user connects to the database, then we can change the text in the "connect" button, to "disconnect", and offer that functionality as well.

Let me know your thoughts.

**Comment by jeleniste** [comment]
@VasiliosKalogirou thanks, but no, I have no time and abilities to lead project, my programming ability is also low. And I`m not to good team player, I need to be managed by somebody. And my main focus is on relational databases (PostGIS).
I think, this is your repository ;), your idea and you have biggest interest (bachelor thesis) to complete it.
You should tell us, what we will do :D.

**Comment by VasiliosKalogirou** [comment]
Ok then, great! I will lead this project if this is ok with @mk45 as well!
@jeleniste I will create a whole new plugin and upload it here. Then I will ask you to add your code as it is because this is your work and then I can take over, and change it accordingly along with @mk45 .
Can I ask you both, when you have some time, to upload a short description of the functions within your plugins? I know this may not sound very nice but whenever you have time you can do it so that I can write the code for the new one based on your code. And also don't forget that probably your old plugins might be removed in favor of the new complete plugin.

**Comment by jeleniste** [comment]
I think, we have to do two main things (classes) 
1.) "something over pymongo" (connect, query, save), only create class able to be connected to gui
2.) mongo<->qgis (memmory) layer?

**Comment by jeleniste** [comment]
I think, that code should be usable straight from commandline, or qgis python console

**Comment by VasiliosKalogirou** [comment]
Why do you suggest having two classes Jan?
I was thinking of one class with several functions (one function for each task). That's why I also asked for you to write me a short description of your functions, so that we can choose which ones to maintain, which ones to merge etc.

**Comment by jeleniste** [comment]
Because i think, that this are two different functionalities
first is inherited from pymongo connection
second from qgis memmory layer

**Comment by VasiliosKalogirou** [comment]
Yes, but what do you mean "first in inherited from pymongo connection and second from qgis memory layer"?

**Comment by jeleniste** [comment]
connect, query, insert... is from pymongo
memmory layer have vizualization and we need add methods for adding data from pymongo (python pickles), it can be recycled for another python libraries able to get data from NOSQL into pickles. And methods, of course, for create pickles (with geometry) from this layer. Maybe, if it is possible, reload data if is changed mapview.

**Comment by VasiliosKalogirou** [comment]
I know about pymongo but I didn't know about memory layer, I 'll have to look at it. If you have any links please send them here so that I research on them and start building the classes on top of them.

**Comment by jeleniste** [comment]
Or we can use ogr qgis layer. It is first decision we should do.

**Comment by jeleniste** [comment]
http://docs.qgis.org/testing/en/docs/pyqgis_developer_cookbook/vector.html

**Comment by jeleniste** [comment]
http://docs.qgis.org/testing/en/docs/pyqgis_developer_cookbook/vector.html#memory-provider

**Comment by VasiliosKalogirou** [comment]
Well, as I understand this is for visualising the data right? This is what you guys have covered with your plugins so what do you consider as the best one to use? 
(Is  "Memory Provider" what you use in your plugin Jan?)

We need to think, do we want to store a shapefile containing the data, or do we want to visualise the data temporarily using the "Memory Provider"?

**Comment by jeleniste** [comment]
You can save layer "as". Yes, I used memory provider for create map layout and atribute table, but maybe is gor provider better.

**Comment by VasiliosKalogirou** [comment]
Or maybe we can ask the user if he wants to save the layer and provide both options?

**Comment by mk45** [comment]
So many subjects I don't know where to start.
1. Having global variables is not good idea.
2. As I said I have deep Python experience, and can smell bad solution/design thats all. @VasiliosKalogirou  You should be project leader. I'm ok with making code. My general advice would be to read documentation twice as deeply and twice as carefully before design something. 
3. I'm ok with removing 'my' plugin in favor of 'ours' under few conditions.
3.1 If I contribute I want to have possibility to access plugin (which would be  also mine then) and fix any further issues, add functionalities and so on (just as I do now). I'm ok with 'We are team' but we act like one then.
3.2 It covers functionality I need.
3.3 Is simple or/and well documented.
4. In Qgis we use classes /inheritance to express some hidden/abstract meaning behind functionality. That makes code readable, usable and just like @jeleniste said even from console. That's why @jeleniste  mentioned two 'obvious' in some way classes (and I do same in my code). One which is 'in fact' modified memory layer (acts like one in terms of how Qgis sees it internally), and second one which connects to db and is plugin's helper class. Because of that, functions inside classes are derivatives of objects idea, and are not worth to be consider, as idea behind my plugin cannot be merged with ex. @jeleniste 's in strait forward way (most probably).
5. It is always good to see bigger picture for such work. Qgis has tools or other plugins which does same for other databases/datasources. If our goal is to make something best/ultimate we must look what and how it is done elsewhere. Ex.: use currently existing ideas/tools like existing edit/save (pencil button/node editing) and so on. What I mean is that I basically would want (and expect it as a user) same from ultimate mongo plugin as now is done for spatialite layers  which is easy connected from file and loaded into Qgis and after edit and unchecking pencil button Qgis asks if to save current layer and saves it! We are IT guys we know options are possible but it may not always be expected by average user. As for me I would make plugin simple as possible (looked like adding postgis or splatialite layer) and things like query are ok but rather for console usage (because Qgis has other tools for selecting objects by ex attribute value)

**Comment by jeleniste** [comment]
My main interest in NOSQL/QGis issues is how to manage "nested" attributes, in qgis is not support for fields like this, is in our ( @mk45 ) posibilities add this functionality?

**Comment by mk45** [comment]
Nested attributes are true power of NoSQL.  Can You push some data example.

**Comment by jeleniste** [comment]
https://raw.githubusercontent.com/jeleniste/mongo_memorizer/master/sampledata/sample.json
This is sample from czech cadastre data, prepared for mongoimport. Some atributes are nested . 

'''json

"ZpusobyOchranyPozemku":{"ZpusobOchrany":{"IdTransakce":0,"Kod":18,"RizeniId":1984709409,"TypOchranyKod":2}}

'''

And can be multiple (arrays). Or I can prepare some more structured dataset from data of forestry management plans. Can be some issue with locales in querying because of interpunction.

In my plugin i create from this pickle json (simplejson.dumps) and save as string, if i want to use it in qgis (for example in symbology), I must write python function for counting attribute in field calculator using simplejson.

**Comment by VasiliosKalogirou** [comment]
In reply to @mk45:
1. - 2. It's obvious you are much more experienced than me and so at some points, I might mention not so good coding examples. Before proceeding to write code I will read on the modules and the documentation and then start work.
3. Definitely, the plugin will be simple, well documented and will cover functionality all of us thinks it's necessary, before being published to the community. Then we can take more ideas on functionality to add.
4. Like I said I will look at the classes and your work in detail before starting coding.
5. For me, the goal would be to create a plugin functioning like DB Manager. If you have experience with this plugin then I think it offers great functionality to connect a database with QGIS, load, query, manipulate and store data. However, at this point, I don't know how to maintain this connection between MongoDB and QGIS and let the user manipulate data real time from QGIS.

**Comment by mk45** [comment]
@jeleniste example you provide might be an easy one. We may have columns names (example only): "ZpusobyOchranyPozemku:ZpusobOchrany:Kod" with value 18. Ugly but works! More problematic would be if You have multiple value (list) in properties ex. "something":[{"athing":5},{"athing":32}]. Did You consider such scenario, or met it somewhere?

@VasiliosKalogirou OK Great

**Comment by jeleniste** [comment]
@mk45 maybe i dont understand, can you get some sample of this data, you think non unique atribute key?

**Comment by mk45** [comment]
@jeleniste Sorry. I If data are like (mongo document example): {"_id":12345, "geometry":{...}, "somedata":"some value", "ZpusobyOchranyPozemku":{"ZpusobOchrany":{ "IdTransakce": 0, "Kod": 18, "RizeniId": 1984709409, "TypOchranyKod": 2}}}. In advance we know that any feature represented in QGIS may have properties binded only as rows in fixed schema table. Schema can have following columns: "_id" , "somedata", "ZpusobyOchranyPozemku:ZpusobOchrany:IdTransakce" , "ZpusobyOchranyPozemku:ZpusobOchrany:RizeniId", "ZpusobyOchranyPozemku:ZpusobOchrany:Kod", ... so on. The ugliest part of this solution is long column names, but is readable and formally correct. 

True power of mongo are nested attributes where one can store multiple additional data (list) in feature. As example we consider postal code addresses: {"_id": 12345 , "geometry":{...}, "postal code":"12-345" , "address":[
{"city":"New Pork",
"street": "Obi-Wan Avenue",
"number": 56},
{"city":"New Pork",
"street": "Long Street",
"number": 34 },
<<many more here>>
]} 
And  for such example, which should also be considered ,because of what mongo allows and is good at, I have no idea how to deal with.

**Comment by jeleniste** [comment]
@mk45 You mean an arrays? ZpusobOchrany means "protection of pracel", you can have more than one of protection on one parcel. For example protection of forest and nature conservation... In my solution is every "nonrealational" fiel converted into json, it doesnt matter if it is in dictionary or in array.

**Comment by jeleniste** [comment]
I have a more complex data i tested, but they are not public, but, i`m able to prepare some dataset from data about czech forestry management.

**Comment by jeleniste** [comment]
Look on attribute ETZ, is that what you mean?

**Comment by mk45** [comment]
@jeleniste Yes arrays. Yes ETZ is good example. Load into json string ... ok I did the same, but it is not user friendly. User can't do  anything with such field nor add to map (as style or size..) nor browse (inline json is hard to read) nor even add it properly to layer (last time I checked QGIS trims too long stings in one column). For 'etz' data maybe better would be to inform (in column value) "long json object",and/or don't add its value at all. Alternatively We may provide some basic aggregate functions (this is an option for further plugin releases). IMHO properties and values should be provide as topmost (not nested).

**Comment by jeleniste** [comment]
@mk45 in my other plugin on forestry data (original is in xml) i`m creating json strings from nested xml fields and I have some custom functions for calculated fields for using in symbology.

**Comment by mk45** [comment]
@jeleniste custom - You mean function created in QGIS (by user). Is it natural/simple to do it this way? Didn't You encounter issues with long json string trim?

**Comment by jeleniste** [comment]
@mk45 Yes, but my functions are created by plugin, we can provide some templates. Natural/simple, I don`t know, import simplejson as json, json.loads... Issues with trim long jsons, maybe it depends on type (length) of field, I haven`t this kind of problem, but json string I had wasn`t really long. I think it will be much better save it in attribute as python pickle, but I`m not able to do this.
